internal/hvsocket/hvsocket.go

```diff
@@ -0,0 +1,81 @@
+//go:build windows
+// +build windows
+
+package hvsocket
+
+import (
+	"context"
+	"fmt"
+	"unsafe"
+
+	"github.com/Microsoft/go-winio/pkg/guid"
+	"golang.org/x/sys/windows"
+
+	"github.com/Microsoft/hcsshim/internal/resources"
+)
+
+const (
+	addressFlagPassthru            = 0x00000001
+	ioCtlHVSocketUpdateAddressInfo = 0x21c004
+)
+
+type addressInfo struct {
+	systemID         guid.GUID
+	virtualMachineID guid.GUID
+	siloID           guid.GUID
+	flags            uint32
+}
+
+type addressInfoCloser struct {
+	handle windows.Handle
+}
+
+var _ resources.ResourceCloser = addressInfoCloser{}
+
+func (aic addressInfoCloser) Release(_ context.Context) error {
+	return windows.CloseHandle(aic.handle)
+}
+
+func CreateAddressInfo(cid, vmid guid.GUID, passthru bool) (resources.ResourceCloser, error) {
+	path := fmt.Sprintf(`\\.\HvSocketSystem\AddressInfo\{%s}`, cid)
+	u16, err := windows.UTF16PtrFromString(path)
+	if err != nil {
+		return nil, err
+	}
+	h, err := windows.CreateFile(
+		u16,
+		windows.GENERIC_READ|windows.GENERIC_WRITE,
+		0,
+		nil,
+		windows.CREATE_NEW,
+		0,
+		0,
+	)
+	if err != nil {
+		return nil, err
+	}
+
+	addrInfo := addressInfo{
```

Comment:
Do we know if this buffer needs to be around until we close the handle created above? (IOCTL handler probably copies this data but it would be good to confirm, otherwise the GC will clean this up and we might end up with a crash).

Reply:
Looks like as long as the input buffer is valid for the duration of the ioctl, we should be good.